incubator/hnc/pkg/forest/forest.go

```diff
@@ -45,6 +67,30 @@ func (f *Forest) Unlock() {
 	f.lock.Unlock()
 }
 
+// AddTypeSyncer adds a reconciler to the types list.
+func (f *Forest) AddTypeSyncer(nss TypeSyncer) {
+	f.types = append(f.types, nss)
+}
+
+// HasTypeSyncer returns true if there is already a reconciler with the given GVK.
+func (f *Forest) HasTypeSyncer(gvk schema.GroupVersionKind) bool {
+	for _, t := range f.types {
+		if t.GetGVK() == gvk {
+			return true
+		}
+	}
+	return false
+}
+
+// GetTypeSyncers returns the types list.
+func (f *Forest) GetTypeSyncers() []TypeSyncer {
+	// We can make a shallow copy here because the TypeSyncer is currently implemented by
+	// *ObjectReconciler (the pointer type) which means we are copying addresses here. We might need
+	// to change this to a deep copy if the TypeSyncer is implemented by a value type in future.
+	types := f.types
```

Comment:
I'm not sure this is true - copying a slice is a shallow operation, see https://play.golang.org/p/huiy48tguSM.

Now, the only think you'll ever do to modify this slice is to append to it, so maybe that's safe? But better safe than sorry, maybe just use `copy` instead: https://stackoverflow.com/questions/30182538/why-cant-i-duplicate-a-slice-with-copy

Reply:
Thanks for the code pointer. Updated the code by following the example.